src/main/java/org/spongepowered/api/event/CauseStackManager.java

```diff
@@ -0,0 +1,168 @@
+     * 
+     * @param obj The object to push to the stack
+     * @return The cause stack manager, for chaining
+     */
+    CauseStackManager pushCause(Object obj);
+
+    /**
+     * Pops the most recently pushed cause object off of the stack and returns
+     * it.
+     * 
+     * @return The last pushed object
+     */
+    Object popCause();
+
+    /**
+     * Pops the most recently <b>n</b> pushed cause objects off of the stack.
+     * 
+     * @param n The number of causes to pop
+     */
+    void popCauses(int n);
+
+    /**
+     * Retrieves but does not remove the most recently pushed cause object.
+     * 
+     * @return The last pushed object
+     */
+    Object peekCause();
+
+    /**
+     * Pushes a frame of the current cause stack and context state.
+     * 
+     * @return A handle for the frame which must be passed back to pop the frame
+     *         from the stack
+     */
+    CauseStackFrame pushCauseFrame();
+
+    /**
+     * Replaces the current cause stack and context with the cause frame at the
+     * top of the frame stack.
+     * 
+     * <p>The frame handle is required to ensure that frames are popped of in
+     * order and are not left in the stack. If an attempt is made to pop a frame
+     * which is not the head of the frame stack then an error will be thrown as
+     * this indicates that a frame was not popped properly.</p>
+     * 
+     * @param handle The frame handle to pop
+     */
+    void popCauseFrame(CauseStackFrame handle);
+
+    /**
+     * Adds the given object to the current context under the given key.
+     * 
+     * @param key The context key
+     * @param value The object
+     * @return The cause stack manager, for chaining
+     * @see EventContextKeys
+     */
+    <T> CauseStackManager addContext(EventContextKey<T> key, T value);
+
+    /**
+     * Gets the context value with the given key.
+     * 
+     * @param key The context key
+     * @return The context object, if present
+     */
+    <T> Optional<T> getContext(EventContextKey<T> key);
+
+    /**
+     * Gets the context value with the given key.
+     * 
+     * <p>If the key is not available, {@link NoSuchElementException} will be
+     * thrown.</p>
+     * 
+     * @param key The context key
+     * @return The context object, if present
+     */
+    default <T> T requireContext(EventContextKey<T> key) {
+        final Optional<T> optional = getContext(key);
+        if (optional.isPresent()) {
+            return optional.get();
+        }
+        throw new NoSuchElementException(String.format("Could not retrieve value for key '%s'", key.getId()));
+    }
+
+    /**
+     * Removes the given context key from the current context.
+     * 
+     * @param key The key to clear
+     * @return The existing context value, if it was present
+     * @see EventContextKeys
+     */
+    <T> Optional<T> removeContext(EventContextKey<T> key);
+
+    public static interface CauseStackFrame extends AutoCloseable {
```

Comment:
In the interest of having a little bit of cleaner API, could the CauseStackFrame get forwarding methods for pushing a cause and getting the current cause? Take the following:

```
                try (final CauseStackManager.StackFrame frame = Sponge.getCauseStackManager().pushCauseFrame()) {
                    Sponge.getCauseStackManager().pushCause(src);
                    ClaimResult
                        claimResult =
                        GriefPreventionPlugin.instance.dataStore.deleteClaim(claim, Sponge.getCauseStackManager().getCurrentCause(), !this.deleteTopLevelClaim);
                    if (!claimResult.successful()) {
                        player.sendMessage(
                            Text.of(TextColors.RED, claimResult.getMessage().orElse(Text.of("Could not delete claim. A plugin has denied it."))));
                        return CommandResult.success();
}
``` java

This is from @gabizou 's update of GP. Just seems odd to jump right back up to Sponge to grab the manager when we have the frame.

Or if anything, a forwarder on the frame to the manager would do.